clang/docs/LanguageExtensions.rst

```diff
@@ -831,6 +832,14 @@ of different sizes and signs is forbidden in binary and ternary builtins.
                                                 semantics, see `LangRef
                                                 <http://llvm.org/docs/LangRef.html#llvm-min-intrinsics-comparation>`_
                                                 for the comparison.
+ T __builtin_elementwise_clz(T x[, T y])        return the number of leading 0 bits in the first argument. If          integer types
+                                                the first argument is 0 and an optional second argument is provided,
+                                                the second argument is returned. If the first argument is 0 but only
+                                                one argument is provided, the behaviour is undefined.
+ T __builtin_elementwise_ctz(T x[, T y])        return the number of trailing 0 bits in the first argument. If         integer types
+                                                the first argument is 0 and an optional second argument is provided,
+                                                the second argument is returned. If the first argument is 0 but only
+                                                one argument is provided, the behaviour is undefined.
```

Comment:
Violations are poison, not instant undefined behavior?

Reply:
> Violations are poison, not instant undefined behavior?

Hmm, that's true. I copied this from the documentation for __builtin_clzg/__builtin_ctzg. Is `poison` something that should be mentioned at the language level, though? There's no mention of poison anywhere else in the document.

> Wondering if we should use the names of the llvm intrinsics. If the arguments and behavior are different from the legacy gcc builtins, we don't need to copy the name

That makes sense to me. Alternatively if they were named `builtin_elementwise_c[lt]zg` then it might indicate they behave identically to *those* builtins. But as it stands the use of `ctz` is misleading as it conflates behaviour with the GCC builtins.

Is this RFC territory?

Reply:
On regular __builtin_clz/ctz the gcc docs state "result is undefined". Not the program is undefined

Reply:
Could do an RFC I suppose. I always found the target selection of behavior absurd with the gcc builtins

Reply:
> On regular __builtin_clz/ctz the gcc docs state "result is undefined". Not the program is undefined

That makes sense, yep. I've updated to use that language here.

Reply:
> Could do an RFC I suppose. I always found the target selection of behavior absurd with the gcc builtins

I've started an RFC: https://discourse.llvm.org/t/rfc-introducing-elementwise-clz-ctz-builtins/85862.

Reply:
As expected, nobody has any strong opinions. Neither do I, to be honest. I think either `__builtin_elementwise_ctz` or `__builtin_elementwise_cttz`. Having `g` doesn't make much sense to me if there are no non-g alternatives.

There are already builtins like `__builtin_elementwise_popcount` which map to `llvm.ctpop` so my inclination is to stick with the GCC names. If you feel strongly about having the intrinsic naming I'll go with that.